// shape-drawing/src/parsePath.js
"use strict";

var actionMap = {
    M: "moveTo",
    L: "lineTo",
    Z: "closePath",
    C: "arcTo",
    Q: "quadraticCurveTo",
    S: "bezierCurveTo"
};

function parsePath(data) {
    if (typeof data === "string") {
        var result = data.match(/\w([\s\d\.,]+\d)?/gi);
        var array = [];
        if (result !== null) {
            array = result.map(function (item) {
                var _result = item.match(/(\w)([\s\d\.,]+)?/);
                if (_result !== null) {
                    var action = actionMap[_result[1].toUpperCase()];
                    if (!action) {
                        throw "invalid path";
                    }
                    return {
                        action: action,
                        data: _result[2] 
                            ? _result[2].split(/[,\s]/).filter(function (item) {return item !== ""}).map(function (item) {return parseFloat(item)})
                            : []
                    };

                }
            });
        }
        return array;
    }

    throw new TypeError("path must be type of string");
}

module.exports = parsePath;
