// shape-drawing/src/shape.js
"use strict";
var currying = require("zz-utils/src/function/currying");
var isPlainObj = require("zz-utils/src/object/isPlainObj");
var apply = require("./apply");
var extendFromCtx = require("./extendFromCtx");
var defaultPointSize = 3;
/**
 * Shape object constructor
 * 
 * @constructor
 * @param {CanvasRenderingContext2D} ctx canvas rendering content
 * @param {Object} options ctx config options
 * @param {Number} options.lineWidth ctx line width
 * @param {DOMString | CanvasGradient | CanvasPattern} options.fillStyle ctx fill style
 * @param {DOMString | CanvasGradient | CanvasPattern} options.strokeStyle ctx stroke style
 * @param {DOMString} options.font ctx font
 * @returns {Shape}
 */
function Shape(ctx, options) {
    this.ctx = ctx;

    if (isPlainObj(options)) {
        var keys = Object.keys(options);
        keys.forEach(function (key) {
            require("./setting")(ctx, key, options[key]);
        })
    }
}

Shape.prototype = Shape.fn = {
    constructor: Shape,
    pointSize: defaultPointSize,
    point: currying(require("./point"), null, defaultPointSize),
    line: currying(require("./line"), null),
    polyline: currying(require("./polyline"), null),
    polygon: currying(require("./polygon"), null),
    circle: currying(require("./circle"), null),
    path: currying(require("./path"), null)
};
Shape.fn.extend = currying(require("./extend"), Shape.fn);

extendFromCtx(Shape.fn);

Shape.prototype.rect = currying(require("./rect"), null);

module.exports = Shape;
